controllers/ocirepository_controller.go

```diff
@@ -942,3 +980,73 @@ func (r *OCIRepositoryReconciler) notify(ctx context.Context,
 		}
 	}
 }
+
+// Verify verifies the authenticity of the given image reference url. First, it tries to keyful approach by looking at
+// whether the given secret exists. Then, if it does not exist, it pushes a keyless approach for verification.
+func (r *OCIRepositoryReconciler) verify(ctx context.Context, obj *sourcev1.OCIRepository, url string) (bool, error) {
+	// get the public keys from the given secret
+	secretRef := obj.Spec.Verify.SecretRef
+
+	// Generate the registry credential keychain either from static credentials or using cloud OIDC
+	keychain, err := r.keychain(ctx, obj)
+	if err != nil {
+		return false, err
+	}
+	authnKeychain := signature.WithAuthnKeychain(keychain)
+
+	ref, err := name.ParseReference(url)
+	if err != nil {
+		return false, err
+	}
+
+	if secretRef != nil {
+		certSecretName := types.NamespacedName{
+			Namespace: obj.Namespace,
+			Name:      secretRef.Name,
+		}
+
+		var pubSecret corev1.Secret
+		if err := r.Get(ctx, certSecretName, &pubSecret); err != nil {
+			return false, err
+		}
+
+		// traverse all public keys and try to verify the signature
+		// this is brute-force approach, but it is ok for now
+		for k, data := range pubSecret.Data {
+			// search for public keys in the secret
+			if strings.HasSuffix(k, ".pub") {
+				verifier, err := signature.New(signature.WithPublicKey(data), authnKeychain)
+				if err != nil {
+					return false, err
+				}
+
+				signatures, _, err := verifier.VerifyImageSignatures(ctx, ref)
+				if err != nil {
+					r.eventLogf(ctx, obj, corev1.EventTypeNormal,
+						sourcev1.SourceVerifiedFailedReason,
+						"failed to verify image %s signature with key %s: %s", url, k, err)
+					continue
+				}
+
+				if signatures != nil {
+					return true, nil
+				}
+			}
+		}
+	} else {
+		verifier, err := signature.New(authnKeychain)
+		if err != nil {
+			return false, err
+		}
+
+		signatures, _, err := verifier.VerifyImageSignatures(ctx, ref)
+		if err != nil {
+			return false, err
+		}
+
+		if len(signatures) > 0 {
+			return true, nil
+		}
+	}
+	return false, nil
```

Comment:
Maybe instead of returning nil here and trying to handle this with if-else in the caller of this function, just returning an error would make the code on the caller's side simpler? A typed error if there's a need to distinguish between the scenarios, which seems to be the same at present in if-else code of the caller.

Reply:
I've updated the code, and there is a chance to address what you are suggesting here, please let me know if I need to fix anything related to the new code, thanks.

Reply:
No longer a concern in the new code. :slightly_smiling_face: